File: db/rust/src/dto/product.rs
//! 商品DTOの定義

use serde::{Deserialize, Serialize};
use utoipa::ToSchema;
use validator::Validate;

/// 商品作成リクエスト
#[derive(Debug, Clone, Serialize, Deserialize, Validate, ToSchema)]
pub struct CreateProductRequest {
    /// 商品コード（16文字以内）
    #[validate(length(min = 1, max = 16))]
    #[schema(example = "PROD00001")]
    pub prod_code: String,

    /// 商品分類コード（8文字以内）
    #[validate(length(max = 8))]
    #[schema(example = "CAT001")]
    pub prod_category_code: Option<String>,

    /// 商品正式名（40文字以内）
    #[validate(length(min = 1, max = 40))]
    #[schema(example = "黒毛和牛サーロインステーキ 200g")]
    pub fullname: String,

    /// 商品略称（10文字以内）
    #[validate(length(min = 1, max = 10))]
    #[schema(example = "サーロイン")]
    pub name: String,

    /// 商品名カナ（20文字以内）
    #[validate(length(min = 1, max = 20))]
    #[schema(example = "クロゲワギュウサーロイン")]
    pub kana: String,

    /// 商品区分
    #[validate(length(max = 1))]
    #[schema(example = "1")]
    pub prod_class: Option<String>,

    /// 製品型番
    #[validate(length(max = 40))]
    pub model_number: Option<String>,

    /// 販売単価
    #[validate(range(min = 0))]
    #[schema(example = 5000)]
    pub unitprice: i32,

    /// 仕入単価
    #[validate(range(min = 0))]
    #[schema(example = 3000)]
    pub purchase_price: Option<i32>,

    /// 売上原価
    #[validate(range(min = 0))]
    #[schema(example = 3500)]
    pub prime_cost: i32,

    /// 税区分（1:課税、2:非課税）
    #[validate(range(min = 1, max = 2))]
    #[schema(example = 1)]
    pub tax_class: Option<i32>,

    /// 在庫管理対象区分
    pub stock_managed: Option<i32>,

    /// 在庫引当区分
    pub stock_reserve: Option<i32>,

    /// 仕入先コード
    #[validate(length(min = 1, max = 8))]
    #[schema(example = "COMP0011")]
    pub sup_code: String,

    /// 仕入先枝番
    #[schema(example = 1)]
    pub sup_seq_num: Option<i32>,
}

impl CreateProductRequest {
    /// ビジネスルールのバリデーション
    pub fn validate_business_rules(&self) -> Result<(), String> {
        if self.unitprice < self.prime_cost {
            return Err("販売単価が売上原価より低い設定はできません".to_string());
        }
        Ok(())
    }
}

/// 商品更新リクエスト（すべてオプション）
#[derive(Debug, Clone, Default, Serialize, Deserialize, Validate, ToSchema)]
pub struct UpdateProductRequest {
    /// 商品分類コード
    #[validate(length(max = 8))]
    pub prod_category_code: Option<String>,

    /// 商品正式名
    #[validate(length(min = 1, max = 40))]
    #[schema(example = "黒毛和牛サーロインステーキ 250g")]
    pub fullname: Option<String>,

    /// 商品略称
    #[validate(length(min = 1, max = 10))]
    pub name: Option<String>,

    /// 商品名カナ
    #[validate(length(min = 1, max = 20))]
    pub kana: Option<String>,

    /// 商品区分
    #[validate(length(max = 1))]
    pub prod_class: Option<String>,

    /// 製品型番
    #[validate(length(max = 40))]
    pub model_number: Option<String>,

    /// 販売単価
    #[validate(range(min = 0))]
    #[schema(example = 5500)]
    pub unitprice: Option<i32>,

    /// 仕入単価
    #[validate(range(min = 0))]
    pub purchase_price: Option<i32>,

    /// 売上原価
    #[validate(range(min = 0))]
    pub prime_cost: Option<i32>,

    /// 税区分
    #[validate(range(min = 1, max = 2))]
    pub tax_class: Option<i32>,

    /// 在庫管理対象区分
    pub stock_managed: Option<i32>,

    /// 在庫引当区分
    pub stock_reserve: Option<i32>,

    /// 仕入先コード
    #[validate(length(min = 1, max = 8))]
    pub sup_code: Option<String>,

    /// 仕入先枝番
    pub sup_seq_num: Option<i32>,
}

/// 商品レスポンス
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct ProductResponse {
    /// 商品コード
    #[schema(example = "PROD00001")]
    pub prod_code: String,

    /// 商品分類コード
    #[schema(example = "CAT001")]
    pub prod_category_code: Option<String>,

    /// 商品正式名
    #[schema(example = "黒毛和牛サーロインステーキ 200g")]
    pub fullname: String,

    /// 商品略称
    #[schema(example = "サーロイン")]
    pub name: String,

    /// 商品名カナ
    #[schema(example = "クロゲワギュウサーロイン")]
    pub kana: String,

    /// 商品区分
    pub prod_class: Option<String>,

    /// 製品型番
    pub model_number: Option<String>,

    /// 販売単価
    #[schema(example = 5000)]
    pub unitprice: i32,

    /// 仕入単価
    pub purchase_price: Option<i32>,

    /// 売上原価
    #[schema(example = 3500)]
    pub prime_cost: i32,

    /// 税区分
    pub tax_class: i32,

    /// 雑区分
    pub misc_class: Option<i32>,

    /// 在庫管理対象区分
    #[schema(example = 1)]
    pub stock_managed: Option<i32>,

    /// 在庫引当区分
    pub stock_reserve: Option<i32>,

    /// 仕入先コード
    #[schema(example = "COMP0011")]
    pub sup_code: String,

    /// 仕入先枝番
    #[schema(example = 1)]
    pub sup_seq_num: Option<i32>,

    /// 作成日時
    pub created_at: chrono::NaiveDateTime,

    /// 更新日時
    pub updated_at: chrono::NaiveDateTime,
}

File: db/go/sales_management/internal/api/schema/product.go
package schema

import "github.com/k2works/sales-management-db/internal/domain"

// CreateProductRequest 商品作成リクエスト
type CreateProductRequest struct {
	ProdCode     string  `json:"prodCode" binding:"required" example:"PROD00001"`
	FullName     string  `json:"fullName" binding:"required" example:"黒毛和牛サーロインステーキ 200g"`
	Name         string  `json:"name" binding:"required" example:"サーロイン"`
	Kana         *string `json:"kana,omitempty" example:"クロゲワギュウサーロイン"`
	UnitPrice    int     `json:"unitPrice" binding:"required,min=0" example:"5000"`
	PoPrice      int     `json:"poPrice" binding:"required,min=0" example:"3500"`
	SupCode      string  `json:"supCode" binding:"required" example:"COMP0011"`
	CategoryCode *string `json:"categoryCode,omitempty" example:"CAT001"`
}

// UpdateProductRequest 商品更新リクエスト（すべてオプション）
type UpdateProductRequest struct {
	FullName     *string `json:"fullName,omitempty" example:"黒毛和牛サーロインステーキ 200g"`
	Name         *string `json:"name,omitempty" example:"サーロイン"`
	Kana         *string `json:"kana,omitempty" example:"クロゲワギュウサーロイン"`
	UnitPrice    *int    `json:"unitPrice,omitempty" example:"5000"`
	PoPrice      *int    `json:"poPrice,omitempty" example:"3500"`
	SupCode      *string `json:"supCode,omitempty" example:"COMP0011"`
	CategoryCode *string `json:"categoryCode,omitempty" example:"CAT001"`
}

// ProductResponse 商品レスポンス
type ProductResponse struct {
	ProdCode     string  `json:"prodCode" example:"PROD00001"`
	FullName     string  `json:"fullName" example:"黒毛和牛サーロインステーキ 200g"`
	Name         string  `json:"name" example:"サーロイン"`
	Kana         *string `json:"kana,omitempty" example:"クロゲワギュウサーロイン"`
	UnitPrice    int     `json:"unitPrice" example:"5000"`
	PoPrice      int     `json:"poPrice" example:"3500"`
	SupCode      string  `json:"supCode" example:"COMP0011"`
	CategoryCode *string `json:"categoryCode,omitempty" example:"CAT001"`
}

// ProductResponseFromDomain ドメインモデルから変換
func ProductResponseFromDomain(product *domain.Product) ProductResponse {
	var supCode string
	if product.SupCode != nil {
		supCode = *product.SupCode
	}

	return ProductResponse{
		ProdCode:     product.ProdCode,
		FullName:     product.FullName,
		Name:         product.Name,
		Kana:         product.Kana,
		UnitPrice:    product.UnitPrice,
		PoPrice:      product.PoPrice,
		SupCode:      supCode,
		CategoryCode: product.CategoryCode,
	}
}

// ErrorResponse エラーレスポンス
type ErrorResponse struct {
	Error   string  `json:"error" example:"エラーメッセージ"`
	Details *string `json:"details,omitempty" example:"詳細情報"`
}
